implementations/elixir/ockam/ockam_hub/lib/hub/service/token_lease_manager.ex

```diff
@@ -1,130 +1,242 @@
+         {:ok, _pid} <- mod.start_link({token_cloud_service, token_cloud_service_address}) do
+      {:ok, mod}
+    else
+      error -> error
+    end
   end
 
-  defp get_all_leases() do
-    # TODO
-    []
+  defp get_cloud_service_module("INFLUXDB"),
+    do: Ockam.TokenLeaseManager.CloudService.Influxdb
+
+  defp get_storage_service_module("POSTGRES"), do: Ockam.TokenLeaseManager.StorageService.Postgres
+
+  defp set_ttl(lease_manager_pid, token_id, ttl) do
+    Process.send_after(lease_manager_pid, {:auto_revoke, token_id}, ttl)
   end
 
-  defp create(cloud_service, options) do
-    {expiration, creation_opts} = Map.pop(options, "expiration")
+  defp do_create(cloud_service, storage_service, options) do
+    {ttl, creation_opts} = Map.pop(options, "ttl")
 
     case cloud_service.create(creation_opts) do
       {:ok, lease} ->
-        save_lease(%Lease{lease | ttl: expiration})
-        set_expiration(self(), lease.id, expiration)
-        {:ok, lease}
+        updated_lease = %Lease{lease | ttl: ttl}
+        storage_service.save(updated_lease)
+        set_ttl(self(), updated_lease.id, ttl)
+        {:ok, updated_lease}
 
       {:error, reason} ->
         {:error, reason}
     end
   end
 
-  defp revoke(cloud_service, token_id) do
+  defp do_revoke(cloud_service, storage_service, token_id) do
     case cloud_service.revoke(token_id) do
-      :ok ->
-        remove_lease(token_id)
-        :ok
-
       {:error, reason} ->
         {:error, reason}
+
+      _other ->
+        storage_service.remove(token_id)
+        :ok
+    end
+  end
+
+  defp do_get(cloud_service, token_id) do
+    case cloud_service.get(token_id) do
+      {:ok, token} -> {:ok, token}
+      :not_found -> {:error, "token #{token_id} does not exist"}
+      error -> error
+    end
+  end
+
+  defp new_ttl(old_ttl, issued_time) when is_integer(old_ttl) and is_binary(issued_time) do
+    case DateTime.from_iso8601(issued_time) do
+      {:ok, issued, _utc_offset} ->
+        diff = DateTime.diff(DateTime.utc_now(), issued, :millisecond)
+        max(old_ttl - diff, 0)
+
+      _error ->
+        0
+    end
+  end
+
+  defp new_ttl(_old_ttl, _issued_time), do: 0
+
+  defp process_encoded_payload(encoded_payload, state) do
+    case decoding_input(encoded_payload) do
+      {:ok, decoded_payload} ->
+        process_decoded_payload(decoded_payload, state)
+
+      error ->
+        encoding_result(error)
+    end
+  end
+
+  defp process_decoded_payload(payload, state) do
+    fun =
+      case payload["action"] do
+        "create" ->
+          fn -> do_create(state[:cloud_service], state[:storage_service], payload["options"]) end
+
+        "revoke" ->
+          fn -> do_revoke(state[:cloud_service], state[:storage_service], payload["token_id"]) end
+
+        "get" ->
+          fn -> do_get(state[:cloud_service], payload["token_id"]) end
+      end
+
+    result = fun.()
+
+    encoding_result(result)
+  end
+
+  defp encoding_result(:ok) do
```

Comment:
Can we use verbs for actions like that (e.g. `encode/decode_result`). `encoding_result` sounds a lot like a noun to me.

Reply:
sure

Reply:
There is one more left: `decoding_input`